### PR TITLE
feat: add a default security policy

### PR DESCRIPTION
**Pull request by fgreinacher**

### CODE REVIEW

**Comment by fgreinacher** [comment]
@jan-kiszka @StefanSchroeder Would you also have a look? This serves as a fallback only. Repo-level security policies are prefered.

File: SECURITY.md
Comment (jan-kiszka) [inline SECURITY.md:13]: Did you intentionally drop sentences like "If the issue is confirmed as a vulnerability by us, we will open a Security Advisory on github and give credits for your report if desired." and "This project follows a 90 day disclosure timeline."? The latter is indeed quite concrete but also not overly challenging.

Comment (StefanSchroeder) [reply SECURITY.md:13]: I have no further comments, except one. Even though 90 days might not be challenging, it will become a metric against which we will be measured... By leaving it out we have more wiggle room. ("we never promised to do this by then.")

**Review by nejch — APPROVED** [review]
Looks nice and simple @fgreinacher.

As I said i think it might be nice to announce this to projects in https://github.com/siemens?q=&type=source before we merge if they don't have `SECURITY.md` / `.github/SECURITY.md` / `docs/SECURITY.md`, since this will become opt-out immediately after merge IIUC. WDYT? Even something silly like this.

```python
import os

from github import Auth, Github

ISSUE_TITLE = "Default Siemens security policy"
ISSUE_DESCRIPTION = """Dear project maintainers,
The default Siemens organization-wide security poolicy will be applied on xx-xx-xxxx.
If you'd like to opt out, please provide your own [`SECURITY.md` policy](https://docs.github.com/en/code-security/getting-started/adding-a-security-policy-to-your-repository).

For more details, see https://github.com/siemens/.github/pull/12."""


auth = Auth.Token(os.getenv("GITHUB_TOKEN"))
gh = Github(auth=auth)
org = gh.get_organization("siemens")

for repo in org.get_repos(type="sources"):
    contents = repo.get_content() + repo.get_content("docs") + repo.get_content(".github")

    if any([f.path.endswith("SECURITY.md") for f in contents]):
        continue

    repo.create_issue(title=ISSUE_TITLE, body=ISSUE_DESCRIPTION)
```

(completely untested so might not work at all but you get teh idea :smile_cat:)

File: SECURITY.md
Comment (fgreinacher) [reply SECURITY.md:13]: @jan-kiszka Yes, I dropped that intentionally because I thought that might be too much to enforce on all maintainers.

But with @nejch suggestion to send a message we could also add it back and tell maintainers to set up their own policy if they want something different.

**Comment by fgreinacher** [comment]
@nejch Great idea with the issue. I'll continue with this after my vacation and we can aim for enabling this in the end of September.

**Comment by jan-kiszka** [comment]
We will also need a way to exclude forks like https://github.com/siemens/linux from this.

**Comment by nejch** [comment]
> We will also need a way to exclude forks like https://github.com/siemens/linux from this.

Maybe we can check that with https://github.com/siemens-testing-main to see if it's applied at all to forks, or if there's a way to disable it. (Edit: seems like it is applied, just checked https://github.com/microsoft/openjdk-jdk/security which has no file).

Otherwise, worst case scenario maybe we can add a note this does not apply to forks with an explicit list? :thinking:

**Comment by jan-kiszka** [comment]
Forks might be used a PR sources, so I'd really like them to be clean, even "just" visually.

**Comment by fgreinacher** [comment]
Hm, that's for sure a blocker. Maybe better to handle that from our internal project, pushing the security policy from there - would also be a bit more explicit and we could notify maintainers there.

**Comment by bufferoverflow** [comment]
I just checked https://github.com/microsoft/openjdk-jdk the only effect is within the About section:

image

There is the `Security policy` which links to the default one, so I do not really see a problem with that. The git repo is of course untouched.

**Comment by nejch** [comment]
@bufferoverflow I was looking at this in the security tab:

https://github.com/microsoft/openjdk-jdk/security

image

But true, it's not in the repo, not sure if that's what @jan-kiszka meant.

**Comment by bufferoverflow** [comment]
Yes, but security tab is no problem from my perspective. We could maybe mention within the policy that this is not valid for forks within the organization.

**Comment by jan-kiszka** [comment]
If there is no file visibly sneaked into the forked repo, just the security tab augmented, a statement that this policy does not apply to forks should be enough.

**Comment by nejch** [comment]
> If there is no file visibly sneaked into the forked repo, just the security tab augmented, a statement that this policy does not apply to forks should be enough.

No, this was never going to be the case :) so this can be solved here in this PR.

**Review by fgreinacher — COMMENTED** [review]
feat: mention scope of security policy

File: SECURITY.md
Comment (fgreinacher) [reply SECURITY.md]: I added back the sentences, but kept the disclosure timeline unspecific. I think this strongly depends on the project and the kind of vulnerability.

WDYT @jan-kiszka?

**Comment by fgreinacher** [comment]
OK, I have the script ready:

```python
import os

from github import Auth, Github
from github.GithubException import GithubException, UnknownObjectException


def get_contents_if_exists(repo, path):
    try:
        return repo.get_contents(path)
    except UnknownObjectException:
        return []


ISSUE_TITLE = "Default Siemens security policy"
ISSUE_DESCRIPTION = """Dear project maintainers,

The default Siemens security policy will be applied to this repository on September 22.

If you'd like to opt out, please [create your own security policy](https://docs.github.com/en/code-security/getting-started/adding-a-security-policy-to-your-repository).

For more details, see the [pull request in our organization repo](https://github.com/siemens/.github/pull/12).

If you have any further questions feel free to comment in the PR or reach out [via email](mailto:opensource@siemens.com).
"""

auth = Auth.Token(os.getenv("GH_TOKEN"))
gh = Github(auth=auth)
org = gh.get_organization(os.getenv("GH_TOKEN"))

for repo in org.get_repos(type="sources"):
    contents = (
        repo.get_contents(".")
        + get_contents_if_exists(repo, "docs")
        + get_contents_if_exists(repo, ".github")
    )

    if any([f.path.endswith("SECURITY.md") for f in contents]):
        continue

    try:
        existing_issue = next(
            iter(
                gh.search_issues(
                    f"is:issue repo:{repo.full_name} in:title {ISSUE_TITLE}"
                )
            ),
            None,
        )
        if existing_issue:
            existing_issue.edit(body=ISSUE_DESCRIPTION)
            print(f"Updated existing issue {existing_issue.html_url}")
        else:
            issue = repo.create_issue(title=ISSUE_TITLE, body=ISSUE_DESCRIPTION)
            print(f"Created issue {issue.html_url}")
    except GithubException as e:
        print(f"Failed to create issue in {repo.html_url}")
```

I'll do the real run with a `siemens-code-bot` PAT.

This it how it looks like: https://github.com/quickest-pint/upgraded-spoon/issues/16.

@nejch @jan-kiszka @r.meier I'd suggest send out the announcement today and merge this on Friday. Any objections?

**Comment by nejch** [comment]
@fgreinacher the issue looks good to me, maybe add bold for the date ;) e.g. 

> will be applied to this repository on **September 22, 2023.**

Friday merge is maybe a bit aggressive for the more dormant repos, but better than waiting for ages IMO. Looking at the actual created issue now, we could also think about auto-closing them after merging this, WDYT?

File: SECURITY.md
Comment (nejch) [inline SECURITY.md:24]: Sorry, missed the comment when I was OoO. The additions look good to me! I know this is copy/pasted but how about:

```suggestion
If we confirm the issue as a vulnerability, we will publish an advisory (e.g. on GitHub) and give credits for your report if desired. We follow the [coordinated vulnerability disclosure](https://vuls.cert.org/confluence/display/CVD) model and will define an appropriate disclosure timeline together with you.
```

**Comment by nejch** [comment]
Ah yes, @fgreinacher that makes sense. I was just thinking about the dozens of issues that may stay open for ages. And maybe it's also a good indicator of stale repositories so we can clean up/archive if needed.

Maybe we can add that they can just close the issue if they're ok with it:

> The default Siemens security policy will be applied to this repository on **September 27, 2023**.
> If you're OK with this, feel free to close this issue.

**Comment by fgreinacher** [comment]
Yeah, let's see how it goes :)

Here is an updated version:

```python
import os

from github import Auth, Github
from github.GithubException import GithubException, UnknownObjectException


def get_contents_if_exists(repo, path):
    try:
        return repo.get_contents(path)
    except UnknownObjectException:
        return []


ISSUE_TITLE = "Default Siemens security policy"
ISSUE_DESCRIPTION = """Dear project maintainers,

The **[default Siemens security policy](https://github.com/siemens/.github/pull/12) will be applied to this repository on September 27, 2023**.

If you're OK with that, you can close this issue. Otherwise, please [create your own security policy](https://docs.github.com/en/code-security/getting-started/adding-a-security-policy-to-your-repository).

If you have any further questions [leave a comment in the pull request](https://github.com/siemens/.github/pull/12) or [send us an email](mailto:opensource@siemens.com).
"""

auth = Auth.Token(os.getenv("GH_TOKEN"))
gh = Github(auth=auth)
org = gh.get_organization(os.getenv("GH_ORG"))

for repo in org.get_repos(type="sources"):
    contents = (
        repo.get_contents(".")
        + get_contents_if_exists(repo, "docs")
        + get_contents_if_exists(repo, ".github")
    )

    if any([f.path.endswith("SECURITY.md") for f in contents]):
        continue

    try:
        existing_issue = next(
            iter(
                gh.search_issues(
                    f"is:issue repo:{repo.full_name} in:title {ISSUE_TITLE}"
                )
            ),
            None,
        )
        if existing_issue:
            existing_issue.edit(body=ISSUE_DESCRIPTION)
            print(f"Updated existing issue {existing_issue.html_url}")
        else:
            issue = repo.create_issue(title=ISSUE_TITLE, body=ISSUE_DESCRIPTION)
            print(f"Created issue {issue.html_url}")
    except GithubException as e:
        print(f"Failed to create issue in {repo.html_url}")
```

Result: https://github.com/quickest-pint/upgraded-spoon/issues/16

File: SECURITY.md
Comment (chombourger) [inline SECURITY.md:8]: typo: organization

File: SECURITY.md
Comment (chombourger) [inline SECURITY.md:17]: etc. usually preceding with a comma foo, bar, etc.

File: SECURITY.md
Comment (chombourger) [inline SECURITY.md:21]: affected version(s)

**Comment by fgreinacher** [comment]
```
Updated existing issue https://github.com/siemens/spdx-licenselist/issues/5
Created issue https://github.com/siemens/tstomp/issues/17
Created issue https://github.com/siemens/android-x86-siemens/issues/1
Failed to create issue in https://github.com/siemens/ASLanPPConnector
Created issue https://github.com/siemens/u-boot/issues/2
Failed to create issue in https://github.com/siemens/codeface
Failed to create issue in https://github.com/siemens/jailhouse
Created issue https://github.com/siemens/dfu-util/issues/1
Failed to create issue in https://github.com/siemens/django-dingos
Failed to create issue in https://github.com/siemens/django-mantis-openioc-importer
Failed to create issue in https://github.com/siemens/django-mantis-stix-importer
Failed to create issue in https://github.com/siemens/django-mantis-core
Failed to create issue in https://github.com/siemens/django-mantis-iodef-importer
Created issue https://github.com/siemens/sparring/issues/2
Created issue https://github.com/siemens/omniproperties/issues/4
Failed to create issue in https://github.com/siemens/xenomai
Created issue https://github.com/siemens/restapidoc/issues/17
Failed to create issue in https://github.com/siemens/django-dingos-authoring
Failed to create issue in https://github.com/siemens/django-mantis-authoring
Created issue https://github.com/siemens/embb/issues/114
Created issue https://github.com/siemens/vanilc/issues/1
Failed to create issue in https://github.com/siemens/django-mantis-actionables
Created issue https://github.com/siemens/freertos-cell/issues/4
Created issue https://github.com/siemens/androit_shmem/issues/1
Failed to create issue in https://github.com/siemens/cityhub-sdk
Created issue https://github.com/siemens/linux-ipipe/issues/1
Created issue https://github.com/siemens/qcsv/issues/1
Created issue https://github.com/siemens/JMiniZinc/issues/36
Created issue https://github.com/siemens/OOCSP/issues/1
Created issue https://github.com/siemens/meta-iot2000/issues/181
Created issue https://github.com/siemens/industrialbenchmark/issues/16
Created issue https://github.com/siemens/confluence-dumper/issues/11
Created issue https://github.com/siemens/policy_search_bb-alpha/issues/1
Created issue https://github.com/siemens/M-Fit/issues/1
Created issue https://github.com/siemens/meta-efibootguard/issues/25
Created issue https://github.com/siemens/ros-sharp/issues/448
Created issue https://github.com/siemens/mtda/issues/339
Failed to create issue in https://github.com/siemens/vpnc-script
Failed to create issue in https://github.com/siemens/jailhouse-images
Created issue https://github.com/siemens/sourcegrid/issues/11
Created issue https://github.com/siemens/ProductConfigurationWithSHACL/issues/1
Created issue https://github.com/siemens/fluffi/issues/279
Created issue https://github.com/siemens/embeddedCMP/issues/2
Created issue https://github.com/siemens/gitlab-test/issues/2
Created issue https://github.com/siemens/pycontainerd/issues/15
Created issue https://github.com/siemens/meta-iot2050/issues/484
Created issue https://github.com/siemens/snap-to-bucket/issues/16
Created issue https://github.com/siemens/siapp-sdk/issues/7
Created issue https://github.com/siemens/k3-rti-wdt/issues/4
Created issue https://github.com/siemens/simatic-ai-launcher/issues/17
Created issue https://github.com/siemens/fact/issues/2
Created issue https://github.com/siemens/libuta/issues/23
Created issue https://github.com/siemens/evaluation-framework/issues/3
Created issue https://github.com/siemens/link-checker-service/issues/20
Created issue https://github.com/siemens/LightweightCmpRa/issues/42
Created issue https://github.com/siemens/opc-ua-pubsub-dotnet/issues/52
Created issue https://github.com/siemens/AzureAD-OAuth-CodeExamples/issues/1
Created issue https://github.com/siemens/GoScans/issues/8
Created issue https://github.com/siemens/Large-Scale-Discovery/issues/4
Created issue https://github.com/siemens/isar-riscv/issues/8
Created issue https://github.com/siemens/ZapSmtp/issues/1
Created issue https://github.com/siemens/dtasm/issues/14
Created issue https://github.com/siemens/libsecutils/issues/55
Created issue https://github.com/siemens/puppet-pam_pkcs11/issues/28
Created issue https://github.com/siemens/powergym/issues/8
Created issue https://github.com/siemens/simatic-s7-webserver-api/issues/33
Created issue https://github.com/siemens/meta-coral/issues/41
Created issue https://github.com/siemens/mkdocs-siemens/issues/2
Created issue https://github.com/siemens/.github/issues/15
Created issue https://github.com/siemens/OOASP/issues/30
Created issue https://github.com/siemens/sicam-gridedge-configurationtemplates/issues/13
Failed to create issue in https://github.com/siemens/ix-docs
Created issue https://github.com/siemens/cyclonedx-property-taxonomy/issues/2
Created issue https://github.com/siemens/continuous-clearing/issues/91
Created issue https://github.com/siemens/pydcc/issues/1
Created issue https://github.com/siemens/sinema-rc-tools/issues/1
```

**Comment by gernot-h** [comment]
> > If there is no file visibly sneaked into the forked repo, just the security tab augmented, a statement that this policy does not apply to forks should be enough.
> 
> No, this was never going to be the case :) so this can be solved here in this PR.

So what should we do with the issues in the forks? Just close them?

@jan-kiszka, does it make sense to have outdated forks like https://github.com/siemens/linux-ipipe at all?

**Comment by nejch** [comment]
> > > If there is no file visibly sneaked into the forked repo, just the security tab augmented, a statement that this policy does not apply to forks should be enough.
> > 
> > 
> > No, this was never going to be the case :) so this can be solved here in this PR.
> 
> So what should we do with the issues in the forks? Just close them?
> 
> @jan-kiszka, does it make sense to have outdated forks like https://github.com/siemens/linux-ipipe at all?

@gernot-h we're already discussing this in our community management repo, I'll ping you there so can join the discussion.

Regarding issues in forks - we actually filtered out forks on GitHub, but `linux-ipipe` is not technically a fork on GitHub as I guess it wasn't forked from the GitHub mirror. IMO it can be closed there and yes, probably archived if it's not active.

File: SECURITY.md
Comment (fgreinacher) [reply SECURITY.md]: Thanks, adapted!

File: SECURITY.md
Comment (fgreinacher) [reply SECURITY.md]: Thanks, adapted!

File: SECURITY.md
Comment (fgreinacher) [reply SECURITY.md]: Thanks, adapted!

**Comment by nejch** [comment]
@fgreinacher should we merge this now? I had a reminder set for this, but we have this in draft still :bow:

**Review by nejch — APPROVED** [review]

**Comment by nejch** [comment]
Thanks @fgreinacher & everyone for  your feedback! LGTM.

**Pull request merged**
